app/Command/Codegen.hs

```diff
@@ -0,0 +1,112 @@
+module Command.Codegen (command) where
+
+import Prelude
+
+import           Command.Common (globWarningOnMisses, printWarningsAndErrors)
+
+import           Control.Applicative (many)
+import           Control.Monad (when, unless)
+
+import qualified Data.Aeson.Internal as A
+import           Data.Aeson.Parser (eitherDecodeWith, json)
+import qualified Data.ByteString.Lazy as BSL
+import           Data.Either (lefts, rights)
+import           Data.Foldable (traverse_)
+import qualified Data.Set as S
+import qualified Data.Map as M
+
+import qualified Language.PureScript as P
+import qualified Language.PureScript.CoreFn as CoreFn
+import qualified Language.PureScript.CoreFn.FromJSON as CoreFn
+
+import qualified Options.Applicative as Opts
+
+import           System.Exit (exitFailure)
+import           System.IO (hPutStr, hPutStrLn, stderr)
+
+data CodegenOptions = CodegenOptions
+  { codegenCoreFnInput :: [FilePath]
+  , codegenJSONErrors :: Bool
+  , codegenOutputDir :: FilePath
+  , codegenSourceMaps :: Bool
+  }
+
+codegen :: CodegenOptions -> IO ()
+codegen CodegenOptions{..} = do
+  inputFiles <- globWarningOnMisses (unless codegenJSONErrors . warnFileTypeNotFound) codegenCoreFnInput
+  when (null inputFiles && not codegenJSONErrors) $ do
+    hPutStr stderr $ unlines
+      [ "purs codegen: No input files."
+      , "Usage: For basic information, try the `--help` option."
+      ]
+    exitFailure
+  mods <- traverse parseCoreFn inputFiles
+
+  let
+    filePathMap =
+      M.fromList $ map ((\m -> (CoreFn.moduleName m, Right $ CoreFn.modulePath m)) . snd) $ rights mods
+
+  unless (null (lefts mods)) $ do
+    traverse_ (hPutStr stderr . formatParseError) $ lefts mods
```

Comment:
Since `left mods` is used twice, perhaps this should be turned into a `let` above?
```haskell
let errList = lefts mods

Also, since `filePathMap` isn't used until after the `unless` block, perhaps it should go below w this block but above the `foreigns <- P.inferForeignMoudles filePathMap` line?